src/python/pants/core/util_rules/lockfile_metadata.py

```diff
@@ -72,18 +74,24 @@ class LockfileMetadata:
     To construct an instance of the most recent concrete subclass, call `LockfileMetadata.new()`.
     """
 
-    _LockfileMetadataSubclass = TypeVar("_LockfileMetadataSubclass", bound="LockfileMetadata")
-
+    # N.B. This __must not__ change between multiple versions of a lockfile metadata for a
+    # particular scope. The delimeter for version 1 must be the delimeter for version N.
+    delimeter: ClassVar[str] = "#"
```

Comment:
I don't think this can safely be a classvar. Python Poetry-generated lockfiles will continue to need to use `#` but PEX use `//` - both should use the same Python class I think, as there's no reason to bump the lockfile header version. 

So instead of a ClassVar, I recommend this being an arg to the relevant methods. Also consider not having a default arg - this seems like something callers should be prompted to think explicitly about.

Reply:
I don't think subclassing works at consumption side though. NB that we use `PythonLockfileMetadata.from_lockfile()`, not `PythonLockfileMetadataV2.from_lockfile()` etc. That is because we don't know beforehand which version we're dealing with.

This would work well if it was `PythonLockfileMetadata.from_lockfile(delimiter="#")` - no coupling to a particular subclass, which is crucial.

Reply:
Don't we then have to duplicate `PexLockfileMetadataV1` with `RequirementsStyleLockfileMetadataV1`, and so on for v2 etc?

Reply:
I don't think it's inconceivable either that the PEX metadata could differ from the pip-requirements metadata.

Reply:
I think I'm not following why the class approach is ideal? I empathesize with wanting to do this via the type system, but it seems to be getting pretty complex with needing to think about MRO and subclass relationships. 

As a function arg, it looks like `PythonLockfileMetadata.from_lockfile(delimiter="#")` at consumption side, and `metadata.add_header_to_lockfile(delimiter="#")` at generation time. That's only two calls. 

And note it's not much worse violating DRY than using classes, where we also would need to update the call-sites exactly twice to point to the correct class. The only way it's worse fwict is repeating the string `#`, rather than using a class name.

 Could we keep it simple with functions?

Reply:
done